### DISCUSSION TITLE
What would be a typical import speed per image file?

### DISCUSSION THREAD

**Discussion by asampal** [photoprism/photoprism#3480] [Q&A]
I'm trying to import a number of JPGs of about 2-4 MB each and it's taking about 5 seconds per image. Is this an expected import speed? It seems rather slow.

This is with PhotoPrism running in Docker on my Asustor 5202T NAS (with 32GB RAM) with two ST12000VN0008 set up in RAID1.

**Comment by lastzero** [comment]
The time (in ms) it took to index a file can be found in the Files tab of the Photo Edit dialog, in the Added field at the bottom.

It depends on many factors, such as your library settings (e.g. thumbnail quality and resolutions), features enabled, and of course your hardware. Your NAS uses an Intel Celeron CPU along with magnetic hard drives, so unfortunately your experience will not be as good as for someone using a mid-range desktop CPU with SSD storage:

- https://docs.photoprism.app/getting-started/troubleshooting/performance/#server-cpu
- https://docs.photoprism.app/getting-started/troubleshooting/performance/#storage

Also note that PhotoPrism already creates many thumbnails during indexing and then uses them to analyze the picture contents with TensorFlow. This is a somewhat heavy process, but it also ensures that you are actually done indexing the file and don't have to wait for further processing later - sometimes for days or months, as we have observed with other apps e.g. Google Photos.

**Reply by lastzero** [reply]
On our public demo, ~1.5s seems a typical INDEXING time for a 3 MB JPEG:

![Screenshot 2023-06-15 at 09 31 25](https://github.com/photoprism/photoprism/assets/301686/141dadfe-4a4b-421c-8558-32443e7a46b4)

**Note that the file will first be MOVED or COPYED if you use the optional IMPORT functionality. This causes additional I/O and can be optimized by using a single mount point for the *originals* and *import* directories, so that only the file system entry needs to be updated. Changing this might not be possible with the App Store version though.**

**Comment by asampal** [comment]
Thanks for these details, @lastzero. I checked, and it seems the files that have been imported so far took between around 3 seconds to 11 seconds. As this seems quite slow, I'm a bit wary about importing the rest before trying to figure out if there's a potential problem (either in the PhotoPrism setup or that of the NAS) that would require starting from scratch.

Do you think running just the DB on an SSD would make a significant difference if the HDD is used for file storage? BTW, I'm using MariaDB 11.0.2 - are there any known performance issues with this version?

I'll try the single mount point advice, though it's hard to believe that for an import of only 5.2K files spread over almost a day, the IO improvement would have made a big difference. I'm using the Docker compose stack, for the app, rather than the app store one, so hopefully that will allow for more tweaking.

Could you describe the single mount point approach a little? I'm not sure I fully understand how things would work wrt the creation of new directories on import and and file copying that would be required to move from the import directory structure to the final one created by PhotoPrism. I'm also interested in the de-duplication aspect of importing so this also adds to my confusion of how things would work.

**Reply by lastzero** [reply]
Yes, only using an SSD for MariaDB (and maybe the storage folder) will work and is a good idea. IOPS are certainly a limiting factor as the mechanical heads can only read or write one position on the disk. So when the database writes, this reduces the read performance for your pictures. Reducing the static thumbnail size to ~1280px instead of 2048px would also help, so less files need to be created, leading to less I/O and lower CPU load.
